org.eclipse.jdt.core.compiler.batch/src/org/eclipse/jdt/internal/compiler/ast/InstanceOfExpression.java

```diff
@@ -154,101 +149,64 @@ public void generateOptimizedBoolean(BlockScope currentScope, CodeStream codeStr
+		codeStream.addVariable(this.secretExpressionValue); // off by 1, but shouldn't matter
+	}
 
 	BranchLabel nextSibling = falseLabel != null ? falseLabel : new BranchLabel(codeStream);
 	codeStream.instance_of(this.type, this.type.resolvedType);
 	codeStream.ifeq(nextSibling);
-	codeStream.load(this.secretInstanceOfPatternExpressionValue);
+	if (this.secretExpressionValue != null) {
+		codeStream.load(this.secretExpressionValue);
+	} else {
+		this.expression.generateCode(currentScope, codeStream, true);
+	}
+
 	if (this.pattern instanceof RecordPattern) {
 		this.pattern.generateOptimizedBoolean(currentScope, codeStream, trueLabel, nextSibling);
-		codeStream.load(this.secretInstanceOfPatternExpressionValue);
-		codeStream.checkcast(this.type, this.type.resolvedType, codeStream.position);
 	} else {
 		codeStream.checkcast(this.type, this.type.resolvedType, codeStream.position);
-		codeStream.dup();
 		codeStream.store(this.elementVariable.binding, false);
 	}
 
-	codeStream.load(this.secretInstanceOfPatternExpressionValue);
-	codeStream.removeVariable(this.secretInstanceOfPatternExpressionValue);
-	codeStream.checkcast(this.type, this.type.resolvedType, codeStream.position);
+	if (this.secretExpressionValue != null) {
+		codeStream.removeVariable(this.secretExpressionValue);
+	}
 
-	if (valueRequired && cst == Constant.NotAConstant) {
+	if (valueRequired) {
 		codeStream.generateImplicitConversion(this.implicitConversion);
 	} else {
 		codeStream.pop();
 	}
 	codeStream.recordPositionsFrom(pc, this.sourceStart);
 
-
-	if ((cst != Constant.NotAConstant) && (cst.typeID() == TypeIds.T_boolean)) {
-		pc = codeStream.position;
-		if (cst.booleanValue() == true) {
-			// constant == true
-			if (valueRequired) {
-				if (falseLabel == null) {
-					// implicit falling through the FALSE case
-					if (trueLabel != null) {
-						codeStream.goto_(trueLabel);
-					}
-				}
+	int position = codeStream.position;
+	if (valueRequired) {
+		if (falseLabel == null) {
+			if (trueLabel != null) {
+				// Implicit falling through the FALSE case
+				codeStream.goto_(trueLabel);
 			}
 		} else {
-			if (valueRequired) {
-				if (falseLabel != null) {
-					// implicit falling through the TRUE case
-					if (trueLabel == null) {
-						codeStream.goto_(falseLabel);
-					}
-				}
-			}
-		}
-		codeStream.recordPositionsFrom(pc, this.sourceStart);
-	} else  {
-		// branching
-		int position = codeStream.position;
-		if (valueRequired) {
-			if (falseLabel == null) {
-				if (trueLabel != null) {
-					// Implicit falling through the FALSE case
-					codeStream.pop2();
-					codeStream.goto_(trueLabel);
-				}
+			if (trueLabel == null) {
+				// Implicit falling through the TRUE case
 			} else {
-				if (trueLabel == null) {
-					// Implicit falling through the TRUE case
-					codeStream.pop2();
-				} else {
-					// No implicit fall through TRUE/FALSE --> should never occur
-				}
+				// No implicit fall through TRUE/FALSE --> should never occur
 			}
```

Comment:
The whole new `else` case is a noop. Is this desired?

Reply:
> The whole new `else` case is a noop. Is this desired?

For now, I have left this in consciously - as the comments there serve to describe the flow. 

This is just scratching the surface. The entire area of pattern code generation is undergoing overhaul. So I will revisit this over and over and will have opportunity to prune